egg/zoo/compo_vs_generalization/bst.py

```diff
@@ -0,0 +1,38 @@
+class BinarySearchTree:
+    def __init__(self, value=None, left=None, right=None):
+        self.value = value
+        self.left = left
+        self.right = right
+
+    def insert(self, value):
+        if self.value is None:
+            self.value = value
+            return
+        if value > self.value:
+            if self.right is None:
+                self.right = BinarySearchTree(value)
+            else:
+                self.right.insert(value)
+        else:
+            if self.left is None:
+                self.left = BinarySearchTree(value)
+            else:
+                self.left.insert(value)
```

Comment:
So this is just an anecdotal comment since I think you should replace this data structure with the native python hashmap (`set()`) but I think a drawback with this implementation is that the tree might not be balanced: if you insert all the elements in decreasing (or increasing) order, the tree will be right (or left) branching and so lookup will be O(n).

Also, keep in mind you are using this data structure with tuples, so you should be careful how python handles comparison with these (for the record I think it is by lexicographic order).

All the more reasons why it's simpler to stick with the native implementation :)